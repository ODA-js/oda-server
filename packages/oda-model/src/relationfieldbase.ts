import {
  FieldBaseInput,
  IFieldBase,
  FieldBase,
  FieldBasePersistence,
  FieldBaseMetaInfo,
  FieldBaseInternal,
  FieldBaseOutput,
  FieldBaseMetaInfoACL,
} from './fieldbase';
import { merge } from 'lodash';
import { Nullable, MetaModelType } from './types';
import { IRelation } from './relation';
import { Internal } from './element';

export interface RelationFieldBaseMetaInfoACL extends FieldBaseMetaInfoACL {
  /** if package allowed to create item of relation */
  create: string[];
  /** if package allowed to delete item of relation */
  delete: string[];
}

export interface RelationFieldBaseMetaInfo<
  P extends RelationFieldBasePersistence
> extends FieldBaseMetaInfo<P> {
  acl: RelationFieldBaseMetaInfoACL;
}

export interface RelationFieldBasePersistence extends FieldBasePersistence {}

export interface RelationFieldBaseInput<
  T extends RelationFieldBaseMetaInfo<P>,
  P extends RelationFieldBasePersistence
> extends FieldBaseInput<T, P> {}

export interface RelationFieldBaseOutput<
  T extends RelationFieldBaseMetaInfo<P>,
  P extends RelationFieldBasePersistence
> extends FieldBaseOutput<T, P> {}

/**
 * relation field definition
 */
export interface IRelationFieldBase<
  T extends RelationFieldBaseMetaInfo<P>,
  I extends FieldBaseInput<T, P>,
  P extends RelationFieldBasePersistence,
  O extends FieldBaseOutput<T, P>
> extends IFieldBase<T, I, P, O> {
  readonly relation: IRelation;
}

export interface RelationFieldBaseInternal extends FieldBaseInternal {
  relation: IRelation;
}

const defaultMetaInfo = {
  acl: {
    read: [],
    update: [],
    create: [],
    delete: [],
  },
};
const defaultInput = { metadata: defaultMetaInfo };

export class RelationFieldBase<
  T extends RelationFieldBaseMetaInfo<P>,
  I extends RelationFieldBaseInput<T, P>,
  S extends RelationFieldBaseInternal,
  P extends RelationFieldBasePersistence,
  O extends FieldBaseOutput<T, P>
> extends FieldBase<T, I, S, P, O> implements IRelationFieldBase<T, I, P, O> {
  public get modelType(): MetaModelType {
    return 'relation-base';
  }

  get relation(): IRelation {
    return this[Internal].relation;
  }

  constructor(init: I) {
    super(merge({}, defaultInput, init));
  }

  public updateWith(obj: Nullable<I>) {
    super.updateWith(obj);
  }

  public toObject(): O {
    return merge({}, super.toObject(), {} as Partial<O>);
  }
  public mergeWith(payload: Nullable<I>) {
    super.mergeWith(payload);
  }
}
