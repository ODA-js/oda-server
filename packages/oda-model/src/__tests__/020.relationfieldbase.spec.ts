import 'jest';
import { RelationFieldBase } from '../relationfieldbase';

describe('RelationFieldBase', () => {
  it('default', () => {
    const res = new RelationFieldBase({ name: 'defaultField' });
    expect(res.toObject()).toMatchSnapshot('default');
  });
});
