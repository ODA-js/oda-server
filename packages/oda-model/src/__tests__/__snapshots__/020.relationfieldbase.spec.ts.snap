// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelationFieldBase default: default 1`] = `
Object {
  "args": Object {},
  "description": undefined,
  "inheritedFrom": undefined,
  "metadata": Object {
    "acl": Object {
      "create": Array [],
      "delete": Array [],
      "read": Array [],
      "update": Array [],
    },
    "persistence": Object {
      "derived": false,
      "identity": false,
      "indexed": false,
      "indexes": Object {},
      "persistent": true,
      "required": false,
    },
  },
  "name": "defaultField",
  "title": undefined,
}
`;
